### PR TITLE
Chromosome view yonah

### PR DESCRIPTION
**Pull request by y330**
For yukthi to look at:
Eplant/views/ChromosomeViewer/Viewer/GeneList.Tsx Line 111

### CODE REVIEW

File: Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx]:
```suggestion
	const [geneticElements, setGeneticElements] = useGeneticElements()
```
All Jotai hooks are used similar to useState, in this case you don't need `useSetGeneticElements()` and should just use the setter provided by `useGeneticElements`

File: Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx]:
```suggestion
```

File: Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx]:
```suggestion
			setGeneticElements([...geneticElements, gene])
```
The reason your state change wasn't immediately triggering a re-render is because React checks for [referential equality](https://blog.bitsrc.io/understanding-referential-equality-in-react-a8fb3769be0) to listen for state change. 

In general it is good to view state objects as immutable and on any update to state, create a new object with the reflected changes. In this case you just need to create a new array copy with the new geneticElement appended to it. @zkdan Correct me if I made any errors here.

File: Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx]:
```suggestion
```

File: Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx]:
```suggestion
	useEffect(() => {
```
Import individual exported members from react whenever possible

**Review by Yukthiw — CHANGES_REQUESTED** [review]
Left a few comments largely related to updating the state of the gene list, will do a more formal review after you're finished with this.

File: Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx]:
```suggestion
      setGeneticElements([...geneticElements, geneticElement])
      setActiveGeneId(geneticElement.id)
```
This will update the state with a new memory reference which will trigger a refresh. With this change you can also remove the `activeGeneId` from the dependency array of the Collections component (I can't think of an instance where the list of genes needs updating on active gene change but I could be wrong on this).

File: Eplant/main.tsx
Comment (Yukthiw) [inline Eplant/main.tsx]: Is there a reason that this is commented out?

File: Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx:43]:
```suggestion
  const [geneticElements, setGeneticElements] = useGeneticElements()
```

All of the Jotai atom hooks work like useStates

File: Eplant/UI/LeftNav/Collections.tsx
Comment (Yukthiw) [inline Eplant/UI/LeftNav/Collections.tsx]:
```suggestion
```

File: Eplant/UI/LeftNav/Collections.tsx
Comment (Yukthiw) [inline Eplant/UI/LeftNav/Collections.tsx]:
```suggestion
```

File: Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx:57]:
```suggestion
  const [geneticElements, setGeneticElements] = useGeneticElements()
  const [collections, setCollections] = useCollections()
```

File: Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx]: Would be a bit easier to read if you moved these helpers into a separate file

File: Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx:44]:
```suggestion
```

Can just use the setter function returned by `useGeneticElements()`

File: Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx]: I think you can just initialize this to an empty array, else it can be a bit confusing to debug in the future where this list contains some dummy data.

File: Eplant/views/ChromosomeViewer/Viewer/Viewer.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/Viewer.tsx:24]: nit: If possible try and use MUI boxes to keep things consistent

File: Eplant/views/ChromosomeViewer/index.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/index.tsx]:
```suggestion
```

No need to initialize this here, can just type the `chromosomeViewData` below

File: Eplant/views/ChromosomeViewer/index.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/index.tsx]:
```suggestion
    const chromosomeViewData: ChromosomeItem[] = await fetch(url)
```

File: Eplant/views/ChromosomeViewer/Viewer/GeneAnnotation.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/GeneAnnotation.tsx]: On my local there seems to be a slight offset in the y-direction of the gene annotation lines:

image

It seems to be zoom agnostic so might just require a small adjustment. Let me know if this is just on my end

**Review by Yukthiw — CHANGES_REQUESTED** [review]
Looks great Yonah! Thanks for doing all of this work. A small comments, largely to do with how state is managed. I noticed thing related to the y-coordinate of the geneAnnotation lines but hopefully this should be a relatively straightforward fix. 

I primarily looked at logic so if @zkdan you could take a look at some of the styling related changes that would be great (although it looks good to me).

File: Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx
Comment (y330) [reply Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx:43]: when i change the geneticelements atom hook to useState format the geneInfoPopup  cannot add genes into the sidebar anymore

File: Eplant/views/ChromosomeViewer/Viewer/Viewer.tsx
Comment (y330) [reply Eplant/views/ChromosomeViewer/Viewer/Viewer.tsx:24]: using Box on line 23 messes up the spacing of the chromosomes

File: Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx
Comment (Yukthiw) [inline Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx:82]: It seems that you are firing off an api call regardless of whether the geneticElement information has been stored already, and this is happening on every chromosome. I think this is logic that should occur only once at the top level component (the Viewer component) and the viewer can pass the relevant GeneItems to each chromosome. Ideally you would also only fire off an api call only for the new gene and not on every gene loaded (maintaining a set of gene ids should make this easy).

Let me know if this makes sense or if you'd like to hop on a quick call sometime to go over this.

**edit**

I'll leave this comment for visibility purposes but on second thought, we can leave this for another PR since we'll need to address the un-needed re-renders as well.

**Comment by Yukthiw** [comment]
@y330 Thanks for addressing most of the comments, just a couple things before we get to merging this.

If you could move the helper functions in Chromosome.tsx to another file ("utils.tsx" or whatever) that would help with readability a lot. 

I made a comment about how we are handling the fetching of gene data, right now we are making a ton of api requests every time a new gene is added. Cutting down on these should improve performance by a lot but I think this can be left for another PR, just leave the comment there for reference.

Finally, before merging run `npx prettier Eplant --write` and push the changes to fix formatting issues (in the future adding the format on save feature to your vscode makes things easy)

File: Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx
Comment (y330) [reply Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx:82]: Yeah, that makes sense, it can definitely be moved to the top level.

**Comment by y330** [comment]
@Yukthiw I moved the helper functions to a new file, and ran npx prettier Eplant --write. I am still not a contributor yet, so i cannot merge the PR

**Review by Yukthiw — APPROVED** [review]
LGTM, thanks Yonah!

**Pull request merged**
